File: Problems 1700-1799/1751. Maximum Number of Events That Can Be Attended II/solution.ts
function bisectRight(events: number[][], start: number, end: number, target: number): number {
    let lo = start;
    let hi = end;

    while (lo < hi) {
        const mid = Math.floor((lo + hi) / 2);

        if (events[mid][0] <= target) {
            lo = mid + 1;
        } else {
            hi = mid;
        }
    }

    return lo;
};

function maxValue(events: number[][], k: number): number {
    const n: number = events.length;
    const dp: number[][] = Array.from(Array(n + 1), () => Array(k + 1).fill(0));
    events.sort((a, b) => a[0] - b[0]);

    for (let i = n - 1; i >= 0; i--) {
        // Left boundary index of events you can attend if you attend events[i]
        const nextStart = bisectRight(events, i + 1, n, events[i][1]);
        for (let count = 1; count <= k; count++) {
            // Don't attend attend or attend event
            dp[i][count] = Math.max(dp[i + 1][count], events[i][2] + dp[nextStart][count - 1]);
        }
    }

    return dp[0][k];
};

File: Problems 1700-1799/1751. Maximum Number of Events That Can Be Attended II/solution.py
class Solution:
    def maxValue(self, events: List[List[int]], k: int) -> int:
        # n = len(events)
        # events.sort(key=lambda x: x[1])
        # ends = [end for _, end, _ in events]
        # dp = [[0] * (k + 1) for _ in range(n + 1)]

        # for i in range(1, n + 1):
        #     prev = bisect.bisect_right(ends, events[i-1][0] - 1)
        #     for count in range(1, k + 1):
        #         dp[i][count] = max(dp[i-1][count], events[i-1][2] + dp[prev][count-1])
    
        # return dp[n][k]

        n = len(events)
        events.sort(key=lambda x: x[0])
        starts = [start for start, _, _ in events]
        dp = [[-1] * (k + 1) for _ in range(n)]

        def solve(pos, k):
            if pos == n or k == 0:
                return 0
            
            if dp[pos][k] != -1:
                return dp[pos][k]

            # nextPos = bisect.bisect_right(events, events[pos][1], lo=pos + 1, key=lambda x: x[0])
            nextPos = bisect_right(starts, events[pos][1])
            
            dp[pos][k] = max(solve(pos + 1, k), events[pos][2] + solve(nextPos, k - 1))
            return dp[pos][k]
        
        return solve(0, k)